waimak-rdbm-ingestion/src/main/scala/com/coxautodata/waimak/rdbm/ingestion/SQLServerExtractor.scala

```diff
@@ -2,33 +2,96 @@ package com.coxautodata.waimak.rdbm.ingestion
+                         , extraConnectionProperties: Properties = new Properties()
+                         , override val transformTableNameForRead: String => String = identity) extends SQLServerBaseExtractor(sqlServerConnectionDetails, extraConnectionProperties) with Logging {
 
   override val sourceDBSystemTimestampFunction = "CURRENT_TIMESTAMP"
 
-  override def escapeKeyword(keyword: String) = s"[$keyword]"
-}
+  val pkQuery: String =
+    s"""(
+       |SELECT p1.schemaName, p1.tableName,
+       |stuff( (SELECT ';'+primaryKeys
+       |	FROM (
+       |		SELECT SCHEMA_NAME(main.schema_id) as schemaName,
+       |		main.name AS tableName,
+       |		tc.name as primarykeys
+       |		FROM sys.tables main
+       |		inner join sys.indexes i on main.object_id = i.object_id
+       |		inner join sys.index_columns ic on i.object_id = ic.object_id and i.index_id = ic.index_id
+       |		inner join sys.columns tc on ic.object_id = tc.object_id and ic.column_id = tc.column_id
+       |		where i.is_primary_key = 1
+       |		group by main.schema_id,
+       |		main.name,
+       |		tc.name) p2
+       |	WHERE p2.tablename = p1.tablename
+       |	ORDER BY primaryKeys
+       | FOR XML PATH(''), TYPE).value('.', 'varchar(max)') ,1,1,'') AS primaryKeys
+       |FROM (
+       |	SELECT SCHEMA_NAME(main.schema_id) as schemaName,
+       |	main.name AS tableName,
+       |	tc.name as primarykeys
+       |	FROM sys.tables main
+       |	inner join sys.indexes i on main.object_id = i.object_id
+       |	inner join sys.index_columns ic on i.object_id = ic.object_id and i.index_id = ic.index_id
+       |	inner join sys.columns tc on ic.object_id = tc.object_id and ic.column_id = tc.column_id
+       |	where i.is_primary_key = 1
+       |	group by main.schema_id,
+       |	main.name,
+       |	tc.name) p1
+       |GROUP BY schemaName,
+       |tableName) m""".stripMargin
+
+  lazy val allTablePKs: Map[String, String] = {
+    import sparkSession.implicits._
+    RDBMIngestionUtils.lowerCaseAll(
+      sparkSession.read
+        .option("driver", driverClass)
+        .jdbc(connectionDetails.jdbcString, pkQuery, connectionProperties)
+    )
+      .as[SQLServerTableMetadata]
+      .collect()
+      .map(metadata => s"${metadata.schemaName}.${metadata.tableName}" -> metadata.primaryKeys).toMap
+  }
+
+
+  override def getTableMetadata(dbSchemaName: String
+                                , tableName: String
+                                , primaryKeys: Option[Seq[String]] = None
+                                , lastUpdatedColumn: Option[String] = None): Try[AuditTableInfo] = {
+    ((primaryKeys, getTablePKs(dbSchemaName, transformTableNameForRead(tableName))) match {
+      case (Some(userPKs), Some(pksFromDB)) if userPKs.sorted != pksFromDB.sorted =>
+        Failure(IncorrectUserPKException(userPKs, pksFromDB))
+      case (Some(userPKs), None) => Success(TableExtractionMetadata(dbSchemaName, tableName, userPKs, lastUpdatedColumn))
```

Comment:
check the coverage report, but I think some of these cases may not be covered by the tests